File: apps/openci_worker_cli/lib/logger.dart
import 'dart:async';
import 'dart:io';
import 'package:uuid/uuid.dart';
import 'package:logging/logging.dart' as dart_logging;
import 'cloud_function_caller.dart';

final _log = dart_logging.Logger('BuildLog');
const _uuid = Uuid();

enum LogLevel { info, warning, error }

class LogEntry {
  final String id;
  final String message;
  final LogLevel level;
  final String timestamp;
  final String? stackTrace;

  LogEntry({
    required this.id,
    required this.message,
    required this.level,
    required this.timestamp,
    this.stackTrace,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'message': message,
        'level': level.name,
        'timestamp': timestamp,
        if (stackTrace != null) 'stackTrace': stackTrace,
      };
}

class _BufferGroup {
  final String buildJobId;
  final String runId;
  List<LogEntry> entries = [];
  Timer? timer;

  _BufferGroup({required this.buildJobId, required this.runId});
}

ApiClient? _apiClient;
final _bufferGroups = <String, _BufferGroup>{};
final _activeWrites = <Future<void>>[];

void setLoggerApiClient(ApiClient apiClient) {
  _apiClient = apiClient;
}

_BufferGroup _getBufferGroup(String buildJobId, String runId) {
  final key = '$buildJobId:$runId';
  return _bufferGroups.putIfAbsent(
    key,
    () => _BufferGroup(buildJobId: buildJobId, runId: runId),
  );
}

const _maxBufferCount = 50;
const _flushInterval = Duration(seconds: 1);
const _maxWriteAttempts = 5;
const _initialRetryDelay = Duration(milliseconds: 500);

Future<void> _sendLogsWithRetry(
  String buildJobId,
  String runId,
  List<LogEntry> logs,
) async {
  final client = _apiClient;
  if (client == null) {
    _log.warning('ApiClient not configured for logging. Logs will be discarded.');
    return;
  }

  final payloadLogs = logs.map((e) => e.toJson()).toList();

  for (var attempt = 1; attempt <= _maxWriteAttempts; attempt++) {
    try {
      await client.appendBuildLogs(
        buildJobId: buildJobId,
        runId: runId,
        logs: payloadLogs,
      );
      return;
    } catch (e) {
      if (attempt == _maxWriteAttempts) {
        _log.warning('[BuildLog] Failed to send bulk logs: $e');
        return;
      }
      final delay = _initialRetryDelay * (1 << (attempt - 1));
      await Future.delayed(delay);
    }
  }
}

void _triggerFlush(_BufferGroup group) {
  group.timer?.cancel();
  group.timer = null;

  if (group.entries.isEmpty) return;

  final logsToSend = List<LogEntry>.from(group.entries);
  group.entries.clear();

  final writeFuture = _sendLogsWithRetry(group.buildJobId, group.runId, logsToSend);
  _activeWrites.add(writeFuture);
  writeFuture.whenComplete(() {
    _activeWrites.remove(writeFuture);
  });
}

Future<void> writeBuildLog(
  String buildJobId,
  String runId,
  LogLevel level,
  String message, {
  String? stackTrace,
}) async {
  final group = _getBufferGroup(buildJobId, runId);
  group.entries.add(LogEntry(
    id: _uuid.v4(),
    message: message,
    level: level,
    timestamp: DateTime.now().toUtc().toIso8601String(),
    stackTrace: stackTrace,
  ));

  if (group.entries.length >= _maxBufferCount) {
    _triggerFlush(group);
  } else {
    group.timer ??= Timer(_flushInterval, () => _triggerFlush(group));
  }
}

Future<void> flushRemainingLogs() async {
  for (final group in _bufferGroups.values) {
    _triggerFlush(group);
  }

  while (_activeWrites.isNotEmpty) {
    await Future.wait(_activeWrites);
  }
}

Future<void> logInfo(
  String buildJobId,
  String runId,
  String message,
) async {
  _log.info(message);
  await writeBuildLog(buildJobId, runId, LogLevel.info, message);
}

Future<void> logWarning(
  String buildJobId,
  String runId,
  String message,
) async {
  _log.warning(message);
  await writeBuildLog(buildJobId, runId, LogLevel.warning, message);
}

Future<void> logError(
  String buildJobId,
  String runId,
  String message, {
  String? stackTrace,
}) async {
  _log.severe(message);
  if (stackTrace != null) {
    _log.severe('Stack trace: $stackTrace');
  }
  await writeBuildLog(
    buildJobId,
    runId,
    LogLevel.error,
    message,
    stackTrace: stackTrace,
  );
}

/// `act` の出力の "[Workflow/Job]" プレフィックスを "[Job]" の短い形式に短縮します。
String stripActPrefix(String line) {
  final pattern = RegExp(r'^\[([^\]]+)\]\s*(.*)$');
  final match = pattern.firstMatch(line);
  if (match != null) {
    final prefixContent = match.group(1) ?? '';
    final msg = match.group(2) ?? '';
    if (msg.isNotEmpty) {
      final slashIndex = prefixContent.lastIndexOf('/');
      final shortJobName = slashIndex != -1
          ? prefixContent.substring(slashIndex + 1).trim()
          : prefixContent.trim();
      return '[$shortJobName] $msg';
    }
  }
  return line;
}

void setupLogging() {
  dart_logging.Logger.root.level = dart_logging.Level.ALL;
  dart_logging.Logger.root.onRecord.listen((record) {
    stdout.writeln('${record.time} [${record.loggerName}] ${record.level.name}: ${record.message}');
    stdout.flush();
  });
}

File: apps/worker_cli_node/src/logger.ts
import { randomUUID } from "node:crypto";
import { appendBuildLogs } from "./firestore.js";

type LogLevel = "info" | "warning" | "error";

interface LogBufferEntry {
  id: string;
  message: string;
  level: LogLevel;
  timestamp: string;
  stackTrace?: string;
}

interface BufferGroup {
  buildJobId: string;
  runId: string;
  entries: LogBufferEntry[];
  timer: NodeJS.Timeout | null;
}

const maxBufferCount = 50;
const flushIntervalMs = 1000;
const maxWriteAttempts = 5;
const initialRetryDelayMs = 500;

const bufferGroups = new Map<string, BufferGroup>();
let activeWritePromises: Promise<void>[] = [];

function getBufferGroup(buildJobId: string, runId: string): BufferGroup {
  const key = `${buildJobId}:${runId}`;
  let group = bufferGroups.get(key);
  if (!group) {
    group = {
      buildJobId,
      runId,
      entries: [],
      timer: null,
    };
    bufferGroups.set(key, group);
  }
  return group;
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function sendLogsWithRetry(
  buildJobId: string,
  runId: string,
  logs: LogBufferEntry[],
): Promise<void> {
  for (let attempt = 1; attempt <= maxWriteAttempts; attempt++) {
    try {
      await appendBuildLogs({
        buildJobId,
        runId,
        logs,
      });
      return;
    } catch (error) {
      if (attempt === maxWriteAttempts) {
        console.warn(`[BuildLog] Failed to send bulk logs: ${String(error)}`);
        return;
      }
      await delay(initialRetryDelayMs * 2 ** (attempt - 1));
    }
  }
}

function triggerFlush(group: BufferGroup): void {
  if (group.timer) {
    clearTimeout(group.timer);
    group.timer = null;
  }

  if (group.entries.length === 0) return;

  const logsToSend = [...group.entries];
  group.entries = []; // バッファを即時クリア

  const writePromise = sendLogsWithRetry(group.buildJobId, group.runId, logsToSend).finally(() => {
    activeWritePromises = activeWritePromises.filter((p) => p !== writePromise);
  });

  activeWritePromises.push(writePromise);
}

async function writeBuildLog(
  buildJobId: string,
  runId: string,
  level: LogLevel,
  message: string,
  stackTrace?: string,
): Promise<void> {
  const group = getBufferGroup(buildJobId, runId);
  group.entries.push({
    id: randomUUID(),
    message,
    level,
    timestamp: new Date().toISOString(),
    ...(stackTrace ? { stackTrace } : {}),
  });

  if (group.entries.length >= maxBufferCount) {
    triggerFlush(group);
  } else if (!group.timer) {
    group.timer = setTimeout(() => {
      triggerFlush(group);
    }, flushIntervalMs);
  }
}

export async function flushLogs(): Promise<void> {
  // 残っているすべてのバッファを強制的にフラッシュ
  for (const group of bufferGroups.values()) {
    triggerFlush(group);
  }

  // 実行中のすべての書き込みが完了するのを待つ
  while (activeWritePromises.length > 0) {
    await Promise.all(activeWritePromises);
  }
}

export async function logInfo(buildJobId: string, runId: string, message: string): Promise<void> {
  console.log(message);
  await writeBuildLog(buildJobId, runId, "info", message);
}

export async function logWarning(
  buildJobId: string,
  runId: string,
  message: string,
): Promise<void> {
  console.warn(message);
  await writeBuildLog(buildJobId, runId, "warning", message);
}

export async function logError(
  buildJobId: string,
  runId: string,
  message: string,
  stackTrace?: string,
): Promise<void> {
  console.error(message);
  if (stackTrace) console.error(stackTrace);
  await writeBuildLog(buildJobId, runId, "error", message, stackTrace);
}
